horovod/_keras/__init__.py

```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
+                            self.gpu_shadow_vars.append(
+                                grad_aggregation_variable)
+                        assert len(self.gpu_shadow_vars) == len(self.grads)
+                    vars_init_op = tf.variables_initializer(
+                        [self.counter, *self.gpu_shadow_vars])
+                    sess.run(vars_init_op)
+
+            def clear_grads():
+                clear_ops_list = []
+                for idx, grad in enumerate(self.gpu_shadow_vars):
+                    grad_aggregation_variable_name = str(idx)
+                    grad_aggregator = tf.get_variable(
+                        grad_aggregation_variable_name)
+                    clear_op = grad_aggregator.assign(
+                        grad_aggregator.initial_value)
+                    clear_ops_list.append(clear_op)
+                return tf.group(*clear_ops_list)
+
+            def aggregate_grads():
+                aggregation_ops_list = []
+                if self._aggregation_frequency > 1:
+                    for idx, grad in enumerate(self.grads):
+                        if self._sparse_as_dense and \
+                                isinstance(grad, tf.IndexedSlices):
+                            grad = tf.convert_to_tensor(grad)
+                        grad_aggregation_variable_name = str(idx)
+                        grad_aggregator = tf.get_variable(
+                            grad_aggregation_variable_name)
+                        update_op = grad_aggregator.assign_add(grad)
+                        aggregation_ops_list.append(update_op)
+                return aggregation_ops_list
+
+            def allreduce_grads():
+                if self._aggregation_frequency > 1:
+                    # Read in latest variables values.
+                    aggregated_grads = []
+                    aggregation_read_ops_list = []
+                    with tf.variable_scope("aggregation_variables", reuse=True):
+                        for idx, grad in enumerate(self.gpu_shadow_vars):
+                            grad_aggregation_variable_name = str(idx)
+                            grad_aggregator = tf.get_variable(
+                                grad_aggregation_variable_name)
+                            aggregated_grads.append(
+                                grad_aggregator.read_value())
+                            aggregation_read_ops_list.append(
+                                aggregated_grads[idx])
+                    aggregation_read_ops = tf.group(
+                        *aggregation_read_ops_list)
+                else:
+                    aggregated_grads = self.grads
+                    aggregation_read_ops = tf.no_op()
+
+                with tf.control_dependencies([aggregation_read_ops]):
+                    averaged_gradients = []
+                    for idx, grad in enumerate(aggregated_grads):
+                        if grad is not None:
```

Comment:
We want to average our locally aggregated gradients by `_aggregation_frequency` prior to calling `hvd.allreduce()`. Or after is fine also. Possibly this could replace the useless `tf.identity()` function we have in the return statement of this function.

Reply:
yeah I like this idea of replacing that `tf.identity()` with the div by `self._aggregation_frequency`